### PR TITLE
Matteo birefnet training

### PR DESCRIPTION
**Pull request by MatteoKartoon**
BiRefNet training finally working

### CODE REVIEW

**Comment by MatteoKartoon** [comment]
Ok

File: .config_batch_size
Comment (joelkartoon) [inline .config_batch_size:1]: remove this file

File: config.py [hunk 1]
```diff
@@ -51,19 +54,24 @@ def __init__(self) -> None:
         self.squeeze_block = ['', 'BasicDecBlk_x1', 'ResBlk_x4', 'ASPP_x3', 'ASPPDeformable_x3'][1]
         self.dec_blk = ['BasicDecBlk', 'ResBlk'][0]
 
-        # TRAINING settings
-        self.batch_size = 4
+        # Try to read batch size from file if it exists
```
Comment (joelkartoon) [inline config.py:57]: fix this

File: config.py [hunk 2]
```diff
@@ -194,10 +204,18 @@ def __init__(self) -> None:
     parser = argparse.ArgumentParser(description='Only choose one argument to activate.')
     parser.add_argument('--print_task', action='store_true', help='print task name')
     parser.add_argument('--print_testsets', action='store_true', help='print validation set')
+    parser.add_argument('--set_batch_size', type=int, help='set the batch size')
     args = parser.parse_args()
 
     config = Config()
+    # Handle the set_batch_size parameter
```
Comment (joelkartoon) [inline config.py:211]: fix this also with previous comment on `.config_batch_size` file

File: config.py [hunk 3]
```diff
@@ -194,10 +204,18 @@ def __init__(self) -> None:
...
+    if hasattr(args, 'set_batch_size') and args.set_batch_size:
+        # Create or modify a config file to store the batch size
+        with open('.config_batch_size', 'w') as f:
+            f.write(str(args.set_batch_size))
+        
+    # Original print functionality
     for arg_name, arg_value in args._get_kwargs():
-        if arg_value:
+        if arg_value and arg_name.startswith('print_'):
```
Comment (joelkartoon) [inline config.py:219]: what does this do?

File: eval_existingOnes.py
Comment (joelkartoon) [inline eval_existingOnes.py:1]: keep original script as a reference, and create your own new script (with a better name)

File: evaluation.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Evaluation
+ckpt_path=$1
+
+log_dir=e_logs && mkdir ${log_dir}
+
+task=$(python3 config.py --print_task)
+
+testset='fine_tuning'
+
+nohup python eval_existingOnes.py --ckpt_path ${ckpt_path} --data_lst ${testset} > ${log_dir}/eval_${testset}.out 2>&1 &
```
Comment (joelkartoon) [inline evaluation.sh:10]: about `nohup` | for later - let's use `tmux` for this, we can set it up

File: split.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+from typing import List
+from datetime import datetime
+import shutil
+from PIL import Image
+import numpy as np
+
+config = Config()
+
+
+def check_121_coupling(train_images: List[str], val_images: List[str], test_images: List[str], original_images: List[str]):
```
Comment (joelkartoon) [inline split.py:14]: nit: `one_to_one`

Comment (joelkartoon) [reply split.py:14]: actually you should have a function that returns a `bool` and call it `check_correct_split`

Comment (joelkartoon) [reply split.py:14]: make sure `original_images` is actually a list of strings (not a list of list of strings), and parse that string in this function

File: split.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    """
+    Print the image information in a special format
+    Args:
+        image: image details vector
+    """
+    return f"Alpha value: {image[0]}, Context: {image[1]}, UUID: {image[2]}, Seed: {image[3]}"
+
+
+
+def split_dataset(train_ratio: float = 0.8, val_ratio: float = 0.1, test_ratio: float = 0.1):
```
Comment (joelkartoon) [inline split.py:135]: should take an argument which is a data path, and another one which is the output folder

File: split.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    # 3. Split according to ratios
+    # 4. Save splits to respective directories
+
+    # Get the folders paths
+    data_folder = config.data_root_dir+"/fine_tuning/"
+    
+    # Get the list of all folders in the data folder
+    for folder in os.listdir(data_folder):
+        #If the folder is a folder of annotated images
+        if folder.startswith("annotated"):
```
Comment (joelkartoon) [inline split.py:158]: avoid this kind of implicit hidden check

File: split.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+            dataset_path_gt = os.path.join(data_folder, folder)
+            
+            #Get the train, test, validation folders paths
+            train_dir=os.path.join(data_folder, os.path.basename(dataset_path_gt).replace("annotated", "train"))
+            val_dir=os.path.join(data_folder, os.path.basename(dataset_path_gt).replace("annotated", "validation"))
+            test_dir=os.path.join(data_folder, os.path.basename(dataset_path_gt).replace("annotated", "test"))
+
+            #Specify the paths of the original images folder
+            dataset_path_im = dataset_path_gt.replace("annotated", "filtred")
```
Comment (joelkartoon) [inline split.py:167]: also change this into an argument

File: split.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+            for f in os.listdir(dataset_path_gt):
+                if f.endswith(".png"):
+                    if save_picture_info(f) in images:
+                        raise ValueError(f"The image {special_print(f)} has already been added to the list")
+                    images.append(save_picture_info(f))
+
+            original_images = [save_picture_info(f) for f in os.listdir(dataset_path_im) if f.endswith('.png')]
+
+            # Shuffle the images to split them randomly
+            random.shuffle(images)
```
Comment (joelkartoon) [inline split.py:188]: as an optional setting to this overall function, give the ability to fix the seed

File: split.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+                os.makedirs(test_dir_im, exist_ok=False)
+                os.makedirs(train_dir_gt, exist_ok=False)
+                os.makedirs(val_dir_gt, exist_ok=False)
+                os.makedirs(test_dir_gt, exist_ok=False)
+            except FileExistsError:
+                print("These directories have already been split")
+                break
+
+            # Save the splitting information
+            with open(os.path.join(data_folder, "splitting.log"), "a") as f:
```
Comment (joelkartoon) [inline split.py:210]: nit, this could be a util function

File: split.sh [hunk 11]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+# Get parameters or use defaults
+train=${1:-0.8}
```
Comment (joelkartoon) [inline split.sh:4]: remove these from bash script - move all this to python (already 99% done)

File: test.sh [hunk 12]
```diff
@@ -1,24 +1,8 @@
-devices=${1:-0}
-pred_root=${2:-e_preds}
+devices="0,1"
+ckpt_path=$1 # e.g. 20250324__1450/epoch_294.pth
```
Comment (joelkartoon) [inline test.sh:2]: adjust this argument directly in the bash script

File: train_test.sh [hunk 13]
```diff
@@ -1,12 +1,12 @@
 #!/bin/sh
-# Example: `setsid nohup ./train_test.sh BiRefNet 0,1,2,3,4,5,6,7 0 &>nohup.log &`
 
-method=${1:-"BSL"}
-devices=${2:-"0,1,2,3,4,5,6,7"}
+method=${1:-"fine_tuning"}
```
Comment (joelkartoon) [inline train_test.sh:3]: nit: set these arguments directly in the bash script

File: visualization.py [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+def compute_interest(scoresss, metrics,many):
+    #Compute an interesting rate for each image based on the difference between the max and min score of the metrics
+    print(f"Computing the most interesting {many} images based on metrics: {metrics}...")
+    interest_values=[]
+    s_copy=copy.deepcopy(scoresss)
+    #Loop through all the images
+    for scoress in s_copy:
+        for ins,scores in enumerate(scoress):
+            scoress[ins] = [scores[i] for i in metrics]
+        print(scoress)
```
Comment (joelkartoon) [inline visualization.py:44]: clean up prints

File: visualization.py [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+
+    # Create a new image with the same size as the original, filled with the green background color
+    green_bg = Image.new("RGBA", original_img.size, (0, 255, 17, 255))
+
+    # Composite the original image onto the green background using the mask
+    result_img = Image.composite(original_img, green_bg, mask_img)
+
+    return result_img
+
+def compute_interest(scoresss, metrics,many):
```
Comment (joelkartoon) [inline visualization.py:35]: nit: maybe a bit over-complex

File: visualization.py [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+        interest_values.append(int_val)
+        print(int_val)
+    print(interest_values)
+    interest_indexes = sorted(range(len(interest_values)), key=lambda i: interest_values[i], reverse=True)[:many]
+    #Return the indexes of the most interesting images
+    return interest_indexes
+
+def do_visualization(model_paths):
+    #function to visualize the results of the models with the scores
+    dm=['S', 'MAE', 'E', 'F', 'WF', 'MBA', 'BIoU', 'MSE', 'HCE']
```
Comment (joelkartoon) [inline visualization.py:55]: investigate metrics more

File: visualization.py [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+    #Create a list of the ground truth and image paths
+    #gt_paths = sorted(glob(os.path.join(config.testsets, 'gt', '*')))
```
Comment (joelkartoon) [inline visualization.py:57]: clean up this part of code with `to_delete`

File: visualization.sh [hunk 18]
```diff
@@ -0,0 +1,6 @@
+# Store parameters in variables
+models_path=$1
```
Comment (joelkartoon) [inline visualization.sh:2]: naming seems off

nit: no need for arguments

File: train.py [hunk 19]
```diff
@@ -138,15 +151,49 @@ def init_models_optimizers(epochs, to_be_distributed):
...
+                class_labels = batch[2]
+            else:
+                inputs = batch[0].to(device)
+                gts = batch[1].to(device)
+                class_labels = batch[2].to(device)
+
+            # Forward pass
+            outputs = trainer.model(inputs)
+            
+            # Handle both training and evaluation modes
```
Comment (joelkartoon) [inline train.py:172]: clean up
specify which part of the code this is drawn from

File: train.py [hunk 20]
```diff
@@ -138,15 +151,49 @@ def init_models_optimizers(epochs, to_be_distributed):
...
+    return val_loss_log.avg  # Return average validation loss
+
 
 class Trainer:
     def __init__(
         self, data_loaders, model_opt_lrsch,
     ):
         self.model, self.optimizer, self.lr_scheduler = model_opt_lrsch
-        self.train_loader = data_loaders
+        self.train_loader = data_loaders[0]
```
Comment (joelkartoon) [inline train.py:191]: please investigate where the `data_loaders` come from & why this is the right thing to do
